### PR TITLE
Bugfixes in simtrack to GEM/CSC digi and stub matching

### PR DESCRIPTION
**Pull request by dildick**
#### PR description:

Follow-up of https://github.com/cms-sw/cmssw/pull/29830. I found a few more bugs while validating the CCLUT algorithm and the GEM trigger primitives in simulation.

#### PR validation:

Tested with Run-3 MC samples in various CMSSW_11_2_X pre-releases.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

N/A

@tahuang1991

### CODE REVIEW

**Comment by dildick** [comment]
@jfernan2 While I'm working on the GEMDigiMatcher, I'll add a section to match SimTracks to GEMDigis based on GEMDigiSimLink (https://github.com/cms-sw/cmssw/pull/30257). The current matching method tends to exclude some digis I've noticed.

**Comment by jfernan2** [comment]
Thanks @dildick 
Just to check if I understand correctly: you plan to add extra commits to this PR?

**Comment by dildick** [comment]
Yes, 2-3 more.

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
+upgrade

**Comment by kpedro88** [comment]
please test

**Comment by jfernan2** [comment]
+1
Despite the important changes in the code, specially in the *SimHitMatchers*, there are no real changes in Validation plots since modules are not run yet

File: Validation/MuonCSCDigis/src/CSCStubMatcher.cc [hunk 1]
```diff
@@ -137,15 +139,28 @@ void CSCStubMatcher::matchCLCTsToSimTrack(const CSCCLCTDigiCollection& clcts) {
       for (const auto& p : comps) {
         layer++;
         for (const auto& q : p) {
+          if (verboseCLCT_)
+            cout << "L" << layer << " " << q << " " << q.getHalfStrip() << " " << std::endl;
```
Comment (kpedro88) [inline Validation/MuonCSCDigis/src/CSCStubMatcher.cc:143]: cout is not allowed, use messagelogger

File: Validation/MuonCSCDigis/src/CSCStubMatcher.cc [hunk 2]
```diff
@@ -137,15 +139,28 @@ void CSCStubMatcher::matchCLCTsToSimTrack(const CSCCLCTDigiCollection& clcts) {
...
           for (const auto& clctComp : (*c).getHits()[layer - 1]) {
-            if (q.getHalfStrip() == clctComp) {
+            if (clctComp == 65535)
+              continue;
+            if (verboseCLCT_) {
+              std::cout << "\t" << clctComp << " " << endl;
```
Comment (kpedro88) [inline Validation/MuonCSCDigis/src/CSCStubMatcher.cc:148]: cout is not allowed

File: Validation/MuonCSCDigis/src/CSCStubMatcher.cc [hunk 3]
```diff
@@ -137,15 +139,28 @@ void CSCStubMatcher::matchCLCTsToSimTrack(const CSCCLCTDigiCollection& clcts) {
...
+            }
+            if (q.getHalfStrip() == clctComp or (isME1a and q.getHalfStrip() + 128 == clctComp)) {
               nMatches++;
+              if (verboseCLCT_) {
+                cout << "\t\tnMatches " << nMatches << std::endl;
```
Comment (kpedro88) [inline Validation/MuonCSCDigis/src/CSCStubMatcher.cc:153]: cout is not allowed

File: Validation/MuonCSCDigis/src/CSCStubMatcher.cc [hunk 4]
```diff
@@ -137,15 +139,28 @@ void CSCStubMatcher::matchCLCTsToSimTrack(const CSCCLCTDigiCollection& clcts) {
...
+              }
             }
           }
         }
+        if (verboseCLCT_) {
+          cout << endl;
```
Comment (kpedro88) [inline Validation/MuonCSCDigis/src/CSCStubMatcher.cc:159]: cout is not allowed

File: Validation/MuonCSCDigis/src/CSCStubMatcher.cc [hunk 5]
```diff
@@ -257,19 +272,23 @@ void CSCStubMatcher::matchLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection& l
...
       bool lct_clct_match(false);
       bool lct_alct_match(false);
       bool lct_gem1_match(false);
       bool lct_gem2_match(false);
 
-      if (verboseLCT_)
-        cout << "in LCT, getCLCT " << lct.getCLCT() << " getALCT " << lct.getALCT() << endl;
-
+      if (verboseLCT_) {
+        cout << ch_id << " " << ch_id2 << endl;
```
Comment (kpedro88) [inline Validation/MuonCSCDigis/src/CSCStubMatcher.cc:281]: cout is not allowed

File: Validation/MuonCSCDigis/src/CSCStubMatcher.cc [hunk 6]
```diff
@@ -257,19 +272,23 @@ void CSCStubMatcher::matchLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection& l
...
+        cout << lct << endl;
+        cout << "getCLCT " << lct.getCLCT() << "\ngetALCT " << lct.getALCT() << "\ngetGEM1 " << lct.getGEM1()
+             << "\ngetGEM2 " << lct.getGEM2() << endl;
+      }
       // Check if matched to an CLCT
       for (const auto& p : clctsInChamber(id)) {
         if (p == lct.getCLCT()) {
           lct_clct_match = true;
+          if (verboseLCT_)
+            cout << "\t...lct_clct_match" << endl;
```
Comment (kpedro88) [inline Validation/MuonCSCDigis/src/CSCStubMatcher.cc:291]: cout is not allowed

File: Validation/MuonCSCDigis/src/CSCStubMatcher.cc [hunk 7]
```diff
@@ -278,6 +297,8 @@ void CSCStubMatcher::matchLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection& l
       for (const auto& p : alctsInChamber(id)) {
         if (p == lct.getALCT()) {
           lct_alct_match = true;
+          if (verboseLCT_)
+            cout << "\t...lct_alct_match" << endl;
```
Comment (kpedro88) [inline Validation/MuonCSCDigis/src/CSCStubMatcher.cc:301]: cout is not allowed

File: Validation/MuonCSCDigis/src/CSCStubMatcher.cc [hunk 8]
```diff
@@ -289,6 +310,8 @@ void CSCStubMatcher::matchLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection& l
         for (const auto& p : gemDigiMatcher_->padsInChamber(gemDetIdL1.rawId())) {
           if (p == lct.getGEM1()) {
             lct_gem1_match = true;
+            if (verboseLCT_)
+              cout << "\t...lct_gem1_match" << endl;
```
Comment (kpedro88) [inline Validation/MuonCSCDigis/src/CSCStubMatcher.cc:314]: cout is not allowed

File: Validation/MuonCSCDigis/src/CSCStubMatcher.cc [hunk 9]
```diff
@@ -298,17 +321,24 @@ void CSCStubMatcher::matchLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection& l
         for (const auto& p : gemDigiMatcher_->padsInChamber(gemDetIdL2.rawId())) {
           if (p == lct.getGEM2()) {
             lct_gem2_match = true;
+            if (verboseLCT_)
+              cout << "\t...lct_gem2_match" << endl;
```
Comment (kpedro88) [inline Validation/MuonCSCDigis/src/CSCStubMatcher.cc:325]: cout is not allowed

File: Validation/MuonCSCDigis/src/CSCStubMatcher.cc [hunk 10]
```diff
@@ -298,17 +321,24 @@ void CSCStubMatcher::matchLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection& l
...
+      const bool clct_gem = lct_clct_match and lct_gem1_match and lct_gem2_match;
+
+      bool lct_tight_matched = alct_clct or alct_gem or clct_gem;
+      bool lct_loose_matched = lct_clct_match or lct_alct_match;
+      bool lct_matched = lct_loose_matched or lct_tight_matched;
 
       if (lct_matched) {
         if (verboseLCT_)
-          cout << "this LCT matched to simtrack in chamber " << ch_id << endl;
+          cout << "...was matched" << endl;
```
Comment (kpedro88) [inline Validation/MuonCSCDigis/src/CSCStubMatcher.cc:341]: cout is not allowed

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 11]
```diff
@@ -55,50 +61,100 @@ void GEMDigiMatcher::match(const SimTrack& t, const SimVertex& v) {
...
+  const edm::DetSetVector<GEMDigiSimLink>& gemDigisSL = *gemDigisSLH_.product();
   const GEMDigiCollection& gemDigis = *gemDigisH_.product();
   const GEMPadDigiCollection& gemPads = *gemPadsH_.product();
   const GEMPadDigiClusterCollection& gemClusters = *gemClustersH_.product();
   const GEMCoPadDigiCollection& gemCoPads = *gemCoPadsH_.product();
 
   clear();
 
+  // hard cut on non-GEM muons
+  if (std::abs(t.momentum().eta()) < 1.55)
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:73]: magic number should be named constant (or even better, obtained from geometry)

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 12]
```diff
@@ -55,50 +61,100 @@ void GEMDigiMatcher::match(const SimTrack& t, const SimVertex& v) {
...
+  matchDigisSLToSimTrack(gemDigisSL);
   matchDigisToSimTrack(gemDigis);
   matchPadsToSimTrack(gemPads);
   matchClustersToSimTrack(gemClusters);
   matchCoPadsToSimTrack(gemCoPads);
 }
 
+void GEMDigiMatcher::matchDigisSLToSimTrack(const edm::DetSetVector<GEMDigiSimLink>& digisSL) {
+  if (verboseSimLink_)
+    cout << "Matching simtrack to GEM simlinks" << endl;
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:86]: cout is not allowed

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 13]
```diff
@@ -55,50 +61,100 @@ void GEMDigiMatcher::match(const SimTrack& t, const SimVertex& v) {
...
+
+  // loop on the simlinks
+  for (auto itsimlink = digisSL.begin(); itsimlink != digisSL.end(); itsimlink++) {
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:89]: range-based loops?

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 14]
```diff
@@ -55,50 +61,100 @@ void GEMDigiMatcher::match(const SimTrack& t, const SimVertex& v) {
...
+      const auto& detids(muonSimHitMatcher_->detIds());
+      if (detids.find(p_id.rawId()) == detids.end())
+        continue;
+
+      // no simhits in this chamber!
+      if (muonSimHitMatcher_->hitsInDetId(p_id.rawId()).empty())
+        continue;
+
+      if (verboseSimLink_)
+        std::cout << "GEMDigiSimLink " << p_id << " " << sl->getStrip() << " " << sl->getBx() << " "
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:103]: cout is not allowed

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 15]
```diff
@@ -55,50 +61,100 @@ void GEMDigiMatcher::match(const SimTrack& t, const SimVertex& v) {
...
+      // loop on the matched simhits
+      for (const auto& simhit : muonSimHitMatcher_->hitsInDetId(p_id.rawId())) {
+        // check if the simhit properties agree
+        if (simhit.particleType() == sl->getParticleType() and simhit.trackId() == sl->getTrackId() and
+            std::abs(simhit.energyLoss() - sl->getEnergyLoss()) < 0.001 and
+            std::abs(simhit.timeOfFlight() - sl->getTimeOfFlight()) < 0.001 and
+            simhit.entryPoint() == sl->getEntryPoint() and simhit.momentumAtEntry() == sl->getMomentumAtEntry()) {
+          detid_to_simLinks_[p_id.rawId()].push_back(*sl);
+          if (verboseSimLink_)
+            cout << "...was matched!" << endl;
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:123]: cout is not allowed

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 16]
```diff
@@ -55,50 +61,100 @@ void GEMDigiMatcher::match(const SimTrack& t, const SimVertex& v) {
...
       // check that the digi is within BX range
       if (d->bx() < minBXDigi_ || d->bx() > maxBXDigi_)
         continue;
-      // check that it matches a strip that was hit by SimHits from our track
-      if (hit_strips.find(d->strip()) == hit_strips.end())
-        continue;
+
       if (verboseDigi_)
-        cout << "...was matched!" << endl;
+        cout << "GEMDigi " << p_id << " " << *d << endl;
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:145]: cout is not allowed

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 17]
```diff
@@ -55,50 +61,100 @@ void GEMDigiMatcher::match(const SimTrack& t, const SimVertex& v) {
...
-      chamber_to_digis_[p_id.chamberId().rawId()].push_back(*d);
-      superchamber_to_digis_[p_id.superChamberId().rawId()].push_back(*d);
+      // check that the digi matches to at least one GEMDigiSimLink
+      for (const auto& sl : detid_to_simLinks_[p_id.rawId()]) {
+        if (sl.getStrip() == d->strip() and sl.getBx() == d->bx()) {
+          detid_to_digis_[p_id.rawId()].push_back(*d);
+          chamber_to_digis_[p_id.chamberId().rawId()].push_back(*d);
+          superchamber_to_digis_[p_id.superChamberId().rawId()].push_back(*d);
+          if (verboseDigi_)
+            cout << "...was matched!" << endl;
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:154]: cout is not allowed

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 18]
```diff
@@ -107,35 +163,28 @@ void GEMDigiMatcher::matchPadsToSimTrack(const GEMPadDigiCollection& pads) {
...
         continue;
+
       if (verbosePad_)
-        cout << "chp1" << endl;
-      // check that it matches a pad that was hit by SimHits from our track
-      if (hit_pads.find(pad->pad()) == hit_pads.end())
-        continue;
-      if (verbosePad_)
-        cout << "chp2" << endl;
+        cout << "GEMPad " << p_id << " " << *pad << endl;
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:175]: cout is not allowed

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 19]
```diff
@@ -107,35 +163,28 @@ void GEMDigiMatcher::matchPadsToSimTrack(const GEMPadDigiCollection& pads) {
...
-      chamber_to_pads_[p_id.chamberId().rawId()].push_back(*pad);
-      superchamber_to_pads_[superch_id()].push_back(*pad);
+      // check that it matches a pad that was hit by SimHits from our track
+      for (auto digi : detid_to_digis_[p_id.rawId()]) {
+        if (digi.strip() / 2 == pad->pad()) {
+          detid_to_pads_[p_id.rawId()].push_back(*pad);
+          chamber_to_pads_[p_id.chamberId().rawId()].push_back(*pad);
+          superchamber_to_pads_[p_id.superChamberId().rawId()].push_back(*pad);
+          if (verbosePad_)
+            cout << "...was matched!" << endl;
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:184]: cout is not allowed

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 20]
```diff
@@ -107,35 +163,28 @@ void GEMDigiMatcher::matchPadsToSimTrack(const GEMPadDigiCollection& pads) {
...
-        continue;
-      if (verbosePad_)
-        cout << "chp2" << endl;
+        cout << "GEMPad " << p_id << " " << *pad << endl;
 
-      detid_to_pads_[id].push_back(*pad);
-      chamber_to_pads_[p_id.chamberId().rawId()].push_back(*pad);
-      superchamber_to_pads_[superch_id()].push_back(*pad);
+      // check that it matches a pad that was hit by SimHits from our track
+      for (auto digi : detid_to_digis_[p_id.rawId()]) {
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:178]: `const auto&`

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 21]
```diff
@@ -144,77 +193,83 @@ void GEMDigiMatcher::matchClustersToSimTrack(const GEMPadDigiClusterCollection&
...
 
     for (auto cluster = clusters_in_det.first; cluster != clusters_in_det.second; ++cluster) {
+      bool isMatched;
+
       // check that the cluster BX is within the range
       if (cluster->bx() < minBXCluster_ || cluster->bx() > maxBXCluster_)
         continue;
 
+      if (verboseCluster_)
+        cout << "GEMCluster " << p_id << " " << *cluster << endl;
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:207]: cout is not allowed

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 22]
```diff
@@ -144,77 +193,83 @@ void GEMDigiMatcher::matchClustersToSimTrack(const GEMPadDigiClusterCollection&
...
+            isMatched = true;
+          }
         }
       }
+      if (isMatched) {
+        detid_to_clusters_[id].push_back(*cluster);
+        chamber_to_clusters_[p_id.chamberId().rawId()].push_back(*cluster);
+        superchamber_to_clusters_[p_id.superChamberId().rawId()].push_back(*cluster);
+        if (verboseCluster_)
+          cout << "...was matched!" << endl;
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:222]: cout is not allowed

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 23]
```diff
@@ -144,77 +193,83 @@ void GEMDigiMatcher::matchClustersToSimTrack(const GEMPadDigiClusterCollection&
...
-      // to match simtrack to GEMCoPad, check the pads within the copad!
-      bool matchL1 = false;
-      GEMDetId gemL1_id(p_id.region(), p_id.ring(), p_id.station(), 1, p_id.chamber(), 0);
       if (verboseCoPad_)
-        cout << "CoPad: chp " << *pad << endl;
-      for (const auto& p : padsInChamber(gemL1_id.rawId())) {
-        if (p == pad->first()) {
-          matchL1 = true;
-          break;
+        cout << "GEMCoPadDigi: " << id << " " << *copad << endl;
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:240]: cout is not allowed

File: Validation/MuonGEMDigis/src/GEMDigiMatcher.cc [hunk 24]
```diff
@@ -144,77 +193,83 @@ void GEMDigiMatcher::matchClustersToSimTrack(const GEMPadDigiClusterCollection&
...
         }
       }
-
-      if (matchL1 and matchL2) {
+      if (isMatchedL1 and isMatchedL2) {
+        superchamber_to_copads_[id.rawId()].push_back(*copad);
         if (verboseCoPad_)
-          cout << "CoPad: was matched! " << endl;
-        superchamber_to_copads_[superch_id()].push_back(*pad);
+          cout << "...was matched! " << endl;
```
Comment (kpedro88) [inline Validation/MuonGEMDigis/src/GEMDigiMatcher.cc:263]: cout is not allowed

File: Validation/MuonHits/src/CSCSimHitMatcher.cc [hunk 25]
```diff
@@ -31,6 +31,12 @@ void CSCSimHitMatcher::match(const SimTrack& track, const SimVertex& vertex) {
   // instantiates the track ids and simhits
   MuonSimHitMatcher::match(track, vertex);
 
+  // hard cut on non-CSC muons
+  if (std::abs(track.momentum().eta()) < 0.9)
```
Comment (kpedro88) [inline Validation/MuonHits/src/CSCSimHitMatcher.cc:35]: magic numbers

File: Validation/MuonHits/src/GEMSimHitMatcher.cc [hunk 26]
```diff
@@ -31,6 +31,10 @@ void GEMSimHitMatcher::match(const SimTrack& track, const SimVertex& vertex) {
   // instantiates the track ids and simhits
   MuonSimHitMatcher::match(track, vertex);
 
+  // hard cut on non-GEM muons
+  if (std::abs(track.momentum().eta()) < 1.55)
```
Comment (kpedro88) [inline Validation/MuonHits/src/GEMSimHitMatcher.cc:35]: magic number

File: Validation/MuonHits/src/GEMSimHitMatcher.cc [hunk 27]
```diff
@@ -69,15 +73,27 @@ void GEMSimHitMatcher::matchSimHitsToSimTrack() {
     for (const auto& h : simHits_) {
       if (h.trackId() != track_id)
         continue;
+
+      const GEMDetId& p_id(h.detUnitId());
+
+      if (verbose_)
+        std::cout << "Candidate GEM simhit " << p_id << " " << h << " " << h.particleType() << " " << h.processType()
```
Comment (kpedro88) [inline Validation/MuonHits/src/GEMSimHitMatcher.cc:80]: cout is not allowed

File: Validation/MuonHits/src/GEMSimHitMatcher.cc [hunk 28]
```diff
@@ -69,15 +73,27 @@ void GEMSimHitMatcher::matchSimHitsToSimTrack() {
...
+
       // discard electron hits in the GEM chambers
-      if (discardEleHits_ && pdgid == 11)
+      if (discardEleHits_ && std::abs(pdgid) == 11)
         continue;
 
-      const GEMDetId& p_id(h.detUnitId());
-      detid_to_hits_[h.detUnitId()].push_back(h);
+      if (verbose_)
+        std::cout << "...was matched" << std::endl;
```
Comment (kpedro88) [inline Validation/MuonHits/src/GEMSimHitMatcher.cc:94]: cout is not allowed

**Comment by dildick** [comment]
@jfernan2 Thanks. Yes, for the time being, the plugin and analyzers are kept in https://github.com/gem-sw/GEMCode/blob/for-CMSSW_11_1_X/GEMValidation/plugins/GEMCSCAnalyzer.cc and https://github.com/gem-sw/GEMCode/tree/for-CMSSW_11_1_X/GEMValidation/interface/Analyzers. I want to move part of it to CMSSW, so that people can do GEM & CSC TP performance studies relatively quickly.

**Comment by kpedro88** [comment]
+upgrade
realized my comments were all in Validation subsystem, which is outside of the upgrade category
(would still be good to address them, but they apply equally to the code as it existed before this PR)

**Comment by jfernan2** [comment]
@dildick could you please implement the changes requested by @kpedro88 in the next iteration of the GEM/CSC Validation?
Thanks

**Comment by dildick** [comment]
@jfernan2 Sure. I can do that.

**Comment by silviodonato** [comment]
+1

**Pull request merged**